pkg/authentication/context.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package authentication
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/oauth2"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+)
+
+type principalContextKey int
+
+const stateSize = 18
+
+var (
+	letterRunes         = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
+	PrincipalContextKey principalContextKey
+)
+
+type Principal struct {
+	Subject string `json:"sub"`
+	Nonce   string `json:"nonce"`
+}
+
+func NewPrincipalFromClaims(c ReadableClaims) (*Principal, error) {
+	a := new(Principal)
+	if err := c.Claims(a); err != nil {
+		return nil, err
+	}
+	return a, nil
+}
+
+func PrincipalContext(ctx context.Context, principal *Principal) context.Context {
+	parent := ctx
+	if ctx == nil {
+		parent = context.Background()
+	}
+
+	return context.WithValue(parent, PrincipalContextKey, principal)
+}
+
+func PrincipalFromContext(ctx context.Context) *Principal {
+	if ctx == nil {
+		return nil
+	}
+
+	return ctx.Value(PrincipalContextKey).(*Principal)
+}
+
+type OIDCProviderSupplier = func(ctx context.Context, issuer string) (*oidc.Provider, error)
+
+type OAuth2Context struct {
+	client   *oauth2.Config
+	verifier TokenVerifier
+
+	tracer  trace.Tracer
+	logger  logging.LoggerInterface
+	monitor monitoring.MonitorInterface
+	rand    *rand.Rand
+}
+
+func (o *OAuth2Context) LoginRedirect(w http.ResponseWriter, r *http.Request) {
+	_, span := o.tracer.Start(r.Context(), "authentication.OAuth2Context.LoginRedirect")
+	defer span.End()
+
+	nonce, _ := nonce(r)
+	state := o.state(stateSize)
```

Comment:
Since cookie support was added for `nonce`, shouldn't we add the same verification logic for `state`?

Reply:
I would honestly skip this for now, since I already adeded that check it should be more than enough for a first implementation. We can add/change that whenever we want. My idea would be to make it part of a CSRF protection implementation we can add in the future.

Reply:
I have no problem not to add this in this PR to unblock development, but we *need* to add state verification eventually. 

Several attacks described in https://datatracker.ietf.org/doc/html/rfc6819 can be mitigated with the `state` parameter.

Reply:
Totally agree on adding state check in the near future.
[Issue](https://github.com/canonical/identity-platform-admin-ui/issues/322) to track this.